### PR TITLE
implement WebSocketStream

### PR DESCRIPTION
**Pull request by KhafraDev**
https://whatpr.org/websockets/48/7b748d3...7b81f79.html#the-websocketstream-interface

Now that the spec has been written, it's time to add it.

```mjs
const ws = new WebSocketStream(`wss://my.websocket.server`)

ws.opened.then((init) => {
  const writer = init.writable.getWriter()

  writer.ready.then(() => writer.write(Buffer.from([65, 66, 67])))
})

```

### CODE REVIEW

**Comment by KhafraDev** [comment]
@nodejs/undici this is ready to review now, although this should not be merged until WebSocketStream is added into the spec. https://github.com/whatwg/websockets/pull/48

**Comment by ronag** [comment]
@KhafraDev fix conflicts? I don't have so much opinion on this one. I'm not interested in the web stuff. Which is why I'd like to separate them out more from undici core.

If you can get one more approving review it should be mergeable. @mcollina

**Comment by KhafraDev** [comment]
Maybe we could start pinging the web standards team for reviews?

File: lib/websocket/symbols.js [hunk 1]
```diff
@@ -8,5 +8,6 @@ module.exports = {
   kBinaryType: Symbol('binary type'),
   kSentClose: Symbol('sent close'),
   kReceivedClose: Symbol('received close'),
-  kByteParser: Symbol('byte parser')
+  kByteParser: Symbol('byte parser'),
+  kPromises: Symbol('kPromises')
```
Comment (Uzlopak) [inline lib/websocket/symbols.js:12]: keep the style?

```suggestion
  kPromises: Symbol('promises')
```

File: types/webidl.d.ts [hunk 2]
```diff
@@ -153,10 +153,15 @@ interface WebidlConverters {
   [Key: string]: (...args: any[]) => unknown
 }
 
+interface WebidlIs {
+  BufferSource(V: any): V is NodeJS.TypedArray | DataView | ArrayBuffer
```
Comment (Uzlopak) [inline types/webidl.d.ts:157]: why any and not unknown?

File: lib/websocket/stream/websocketstream.js [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+      cancel: (reason) => this.#cancel(reason)
+    }, new ByteLengthQueuingStrategy({ highWaterMark: 16384 }))
+
+    // 9.  Let writeAlgorithm be an action that writes chunk to stream , given chunk .
+    // 10.  Let closeAlgorithm be an action that closes stream .
+    // 11.  Let abortAlgorithm be an action that aborts stream with reason , given reason .
+    // 12.  Let writable be a new WritableStream .
+    // 13.  Set up writable with writeAlgorithm , closeAlgorithm , and abortAlgorithm .
+    const writable = new WritableStream({
+      write: (chunk) => this.#write(chunk),
```
[the remark this answers is not in the document]
Comment (KhafraDev) [reply lib/websocket/stream/websocketstream.js:227]: WritableStream should have backpressure built into it.

[the remark this answers is not in the document]
Comment (mcollina) [reply lib/websocket/stream/websocketstream.js:227]: Yes, but you are not forwarding the information of the underlining socket being busy up.
https://github.com/nodejs/undici/pull/2713/files#diff-c1287dec52a96e140d7231c268bfa35a6f1e0390c46ccaf15b3ba5834f2a7342R322 will return `false` if the socket is "full", and then emit a `'drain'` event when free again. This is not connected in any way to the `WritableStream`.

**Pull request closed without merging**